File: proyecto/frontend/src/app/services/factura.service.ts
import { inject, Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Factura, PageResponse, EstadoResumen } from '../models/factura.model';

/**
 * Servicio para consumir los endpoints REST del microservicio Centralizador.
 */
@Injectable({ providedIn: 'root' })
export class FacturaService {

  private http = inject(HttpClient);
  private readonly baseUrl = '/api/facturas';

  /**
   * Lista facturas de forma paginada.
   * @param page número de página (0-indexed)
   * @param size cantidad de registros por página
   * @returns Observable con la página de facturas
   */
  listar(page = 0, size = 20): Observable<PageResponse<Factura>> {
    const params = new HttpParams()
      .set('page', page)
      .set('size', size);
    return this.http.get<PageResponse<Factura>>(this.baseUrl, { params });
  }

  /**
   * Obtiene el resumen de conteo de facturas por estado.
   * @returns Observable con mapa {PENDIENTE: N, PROCESO: N, ...}
   */
  resumenEstados(): Observable<EstadoResumen> {
    return this.http.get<EstadoResumen>(`${this.baseUrl}/estados/resumen`);
  }

  /**
   * Crea facturas en lote.
   * @param facturas lista de facturas a crear
   * @returns Observable con las facturas creadas
   */
  crearEnLote(facturas: Partial<Factura>[]): Observable<Factura[]> {
    return this.http.post<Factura[]>(`${this.baseUrl}/bulk`, facturas);
  }
}

File: proyecto/backend/src/main/java/com/hackathon/centralizador/service/FacturaService.java
package com.hackathon.centralizador.service;

import com.hackathon.centralizador.dto.FacturaRequest;
import com.hackathon.centralizador.dto.FacturaResponse;
import com.hackathon.centralizador.model.EstadoFactura;
import com.hackathon.centralizador.model.Factura;
import com.hackathon.centralizador.repository.FacturaRepository;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Arrays;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * Servicio para operaciones de negocio sobre facturas.
 */
@Service
public class FacturaService {

    private final FacturaRepository facturaRepository;

    public FacturaService(FacturaRepository facturaRepository) {
        this.facturaRepository = facturaRepository;
    }

    /**
     * Crea un lote de facturas a partir de los requests recibidos.
     * Cada factura se persiste con estado PENDIENTE.
     *
     * @param requests lista de DTOs de entrada
     * @return lista de DTOs de salida con id, estado y fechaCreacion asignados
     */
    @Transactional
    public List<FacturaResponse> crearFacturasEnLote(List<FacturaRequest> requests) {
        List<Factura> facturas = requests.stream()
                .map(this::mapToEntity)
                .toList();

        List<Factura> saved = facturaRepository.saveAll(facturas);

        return saved.stream()
                .map(this::mapToResponse)
                .toList();
    }

    /**
     * Lista facturas de forma paginada.
     *
     * @param pageable parámetros de paginación y ordenamiento
     * @return página de facturas mapeadas a DTO
     */
    @Transactional(readOnly = true)
    public Page<FacturaResponse> listar(Pageable pageable) {
        return facturaRepository.findAll(pageable)
                .map(this::mapToResponse);
    }

    /**
     * Retorna un resumen con el conteo de facturas por estado.
     * Incluye todos los estados del enum aunque tengan conteo 0.
     *
     * @return mapa con cada estado y su conteo
     */
    @Transactional(readOnly = true)
    public Map<String, Long> resumenEstados() {
        Map<String, Long> resumen = Arrays.stream(EstadoFactura.values())
                .collect(Collectors.toMap(
                        Enum::name,
                        estado -> 0L,
                        (a, b) -> a,
                        LinkedHashMap::new
                ));

        facturaRepository.contarPorEstado().forEach(row -> {
            EstadoFactura estado = (EstadoFactura) row[0];
            Long count = (Long) row[1];
            resumen.put(estado.name(), count);
        });

        return resumen;
    }

    private Factura mapToEntity(FacturaRequest request) {
        return Factura.builder()
                .cus(request.getCus())
                .fechaPago(request.getFechaPago())
                .responsabilidadFiscal(request.getResponsabilidadFiscal())
                .tipoDocumento(request.getTipoDocumento())
                .numeroDocumento(request.getNumeroDocumento())
                .nombres(request.getNombres())
                .apellidos(request.getApellidos())
                .telefono(request.getTelefono())
                .municipio(request.getMunicipio())
                .direccion(request.getDireccion())
                .email(request.getEmail())
                .valorPack(request.getValorPack())
                .valorPackIva(request.getValorPackIva())
                .comentarios(request.getComentarios())
                .estado(EstadoFactura.PENDIENTE)
                .build();
    }

    private FacturaResponse mapToResponse(Factura factura) {
        return FacturaResponse.builder()
                .id(factura.getId())
                .cus(factura.getCus())
                .fechaPago(factura.getFechaPago())
                .responsabilidadFiscal(factura.getResponsabilidadFiscal())
                .tipoDocumento(factura.getTipoDocumento())
                .numeroDocumento(factura.getNumeroDocumento())
                .nombres(factura.getNombres())
                .apellidos(factura.getApellidos())
                .telefono(factura.getTelefono())
                .municipio(factura.getMunicipio())
                .direccion(factura.getDireccion())
                .email(factura.getEmail())
                .valorPack(factura.getValorPack())
                .valorPackIva(factura.getValorPackIva())
                .comentarios(factura.getComentarios())
                .estado(factura.getEstado().name())
                .fechaCreacion(factura.getFechaCreacion())
                .build();
    }
}
